### DISCUSSION TITLE
Correct way to resolve URL objects to paths, e.g. for use in `og:url` tag

### DISCUSSION THREAD

**Discussion by BigglesZX** [vercel/next.js#22025] [Help]
I'm still new to Next.js, so apologies in advance if I'm missing something obvious.

I use `next/link` with dynamic routing in my project – here's an example. I'm using literals here but in practice these are generated from CMS data, but the point is the same.

```jsx
<Link href={{
    pathname: '/work/[slug]',
    query: { 
        slug: 'project-foo',
    },
}}>
    Project Foo
</Link>
```

This works fine, but I have just come across a use-case where I need to output what the above URL object actually **resolves to,** i.e. `/work/project-foo`, as I want to output it as (part of) the value for an `og:url` meta tag.

Upon learning of the existence of the node.js core `url` module and its [in-browser counterpart](https://www.npmjs.com/package/url), I tried using `url.resolve`:

```javascript
import { resolve } from 'url';

resolve({
    pathname: '/work/[slug]',
    query: { 
        slug: 'project-foo',
    },
});
```

However this produces `/work/[slug]?slug=project-foo`. The `url` package doesn't know about Next.js' dynamic routing – fair enough.

I then learned that Next.js implements [its own version](https://github.com/vercel/next.js/blob/canary/packages/next/next-server/lib/router/utils/format-url.ts), so I tried that:

```javascript
import { formatUrl } from 'next/dist/next-server/lib/router/utils/format-url';

formatUrl({
    pathname: '/work/[slug]',
    query: { 
        slug: 'project-foo',
    },
});
```

Unfortunately this produces the same result (`/work/[slug]?slug=project-foo`), not seeming to parse the dynamic parameter placeholder.

Reading the source some more, I tried using [`resolveHref`](https://github.com/vercel/next.js/blob/6b99bda3e3971831c0997ce59b19cb848e4099ff/packages/next/next-server/lib/router/router.ts#L259) from next-server's router package, as I'd seen it used by `next/link` [here](https://github.com/vercel/next.js/blob/6b99bda3e3971831c0997ce59b19cb848e4099ff/packages/next/client/link.tsx#L225):

```javascript
import { resolveHref } from 'next/dist/next-server/lib/router/router';

const [resolvedHref, resolvedAs] = resolveHref(
    '/',
    {
        pathname: '/work/[slug]',
        query: { 
            slug: 'project-foo',
        },
    },
    true
);
```

This gets the result I'm after: `/work/project-foo`. Great!

However:

1. I'm aware that `resolveHref` ideally wants that first argument to come from the router state (that's what `next/link` does), so what I've done here might not be bulletproof, but I don't yet understand if/why that's the case (though maybe this is fine).
2. The deep import from inside the next package makes me a little uncomfortable as I'm clearly not using a well-published top-level utility function and future versions of Next.js might well change this (though maybe this is fine).
3. Overall I've got the feeling I'm probably missing some more obvious/recommended API for getting what I want in this case.

Any thoughts or pointers would be much appreciated. Thanks in advance!

**Comment by BigglesZX** [comment]
Hello, it's me from the future. I'm still using the above, but upgrading to Next 11 (and 12) required me to change the import of `resolveHref` to:

```js
import { resolveHref } from 'next/dist/shared/lib/router/router';
```

**Comment by BigglesZX** [comment]
Hello, it's me from slightly further in the future. In Next 12 the `resolveHref` method signature has changed and it requires a `Router` instance to be passed in – this seems to work with a straightforward import from `next/router`. I'm now using the following code to resolve routes:

```javascript
import Router from 'next/router';
import { resolveHref } from 'next/dist/shared/lib/router/router';

const exampleRoute = {
    pathname: '/work/[slug]',
    query: { 
        slug: 'project-foo',
    },
};
const [, resolvedAs] = resolveHref(Router, exampleRoute, true);
//       ^-- resulting path as a string
```

**Comment by madebyfabian** [comment]
Thanks for noting these changes here 😄. This absolutely should be included in the official package.
I tried it for Next.js 13 but `resolveHref` is no longer exported. The export path has changed to

```ts
import { resolveHref } from 'next/dist/shared/lib/router/utils/resolve-href'
// ...
```

**Reply by geekysaurabh001** [reply]
Any further updates? Lol. It has removed resolveHref it seems.

**Reply by BigglesZX** [reply]
Looks like it's moved again, the source is in [this file](https://github.com/vercel/next.js/blob/canary/packages/next/src/client/resolve-href.ts#L16-L30). I haven't tested this (could someone else confirm?) but the import should be:

```js
import { resolveHref } from 'next/dist/client/resolve-href';
// ...
```

Would be really great if this function could be made "official" so we don't have to chase it around the package on every major release.

**Reply by madebyfabian** [reply]
Indeed, feels like a game of catch every release lol. Should definitely be included officially.
With Nuxt for example, you get an official resolver available + automatically generated route names + typed routes.
![Bildschirmfoto 2023-10-04 um 11 18 10](https://github.com/vercel/next.js/assets/32813692/2d65385f-7519-428f-aedc-9c6ccbc99b4e)

**Reply by geekysaurabh001** [reply]
Tested it now. It is working for my use case of complete URL as it worked previously.

**Comment by szamanr** [comment]
**Next.js 14**
_this is posted in the thread above, but reposting here to make it easier to find_
```
import { resolveHref } from 'next/dist/client/resolve-href';
```
